sdk/python/feast/type_map.py

```diff
@@ -180,6 +183,28 @@ def python_type_to_feast_value_type(
     return type_map[value.dtype.__str__()]
 
 
+def python_values_to_feast_value_type(name: str, values: Any, recurse: bool = True):
+    inferred_dtype = ValueType.UNKNOWN
+    for row in values:
+        current_dtype = python_type_to_feast_value_type(
+            name, value=row, recurse=recurse
+        )
+
+        if inferred_dtype is ValueType.UNKNOWN:
+            inferred_dtype = current_dtype
+        else:
+            if current_dtype != inferred_dtype:
```

Comment:
Doesn't matter too much right now since you call it from the inference of entity rows, which shouldn't have null values, but for feature inference from datasources we'd probably want a different treatment.

Reply:
Huh, you're right that's an interesting edge case I hadn't considered and the behaviour here is a little odd actually because the order matters.

* `[None, None, None, 1]` would correctly be inferred to `ValueType.INT64_LIST`
* `[None, None, 1, None]` would end up being `ValueType.UNKNOWN`

Something to consider 😕 But to me this seems like a very odd edge case. I can't think of that many use cases for a mix of `null` and `not null` values inside the list.

Can you think of an example?

Reply:
I think I've addressed this now - let me know what you think.